src/content/peerconnection/endtoend-encryption/js/worker.js

```diff
@@ -0,0 +1,119 @@
+      'len=' + chunk.data.byteLength,
+      'type=' + (chunk.type || 'audio'),
+      'ts=' + chunk.timestamp,
+      'ssrc=' + chunk.synchronizationSource
+  );
+}
+
+let scount = 0;
+function encodeFunction(chunk, controller) {
+  if (scount++ < 30) { // dump the first 30 packets.
+    dump(chunk, 'send');
+  }
+  if (currentCryptoKey) {
+    const view = new DataView(chunk.data);
+    // Any length that is needed can be used for the new buffer.
+    const newData = new ArrayBuffer(chunk.data.byteLength + 5);
+    const newView = new DataView(newData);
+
+    const cryptoOffset = useCryptoOffset? frameTypeToCryptoOffset[chunk.type] : 0;
+    for (let i = 0; i < cryptoOffset && i < chunk.data.byteLength; ++i) {
+      newView.setInt8(i, view.getInt8(i));
+    }
+    for (let i = cryptoOffset; i < chunk.data.byteLength; ++i) {
+      const keyByte = currentCryptoKey.charCodeAt(i % currentCryptoKey.length);
+      newView.setInt8(i, view.getInt8(i) ^ keyByte);
+    }
+    // Append keyIdentifier.
+    newView.setUint8(chunk.data.byteLength, currentKeyIdentifier % 0xff);
+    // Append checksum
+    newView.setUint32(chunk.data.byteLength + 1, 0xDEADBEEF);
+
+    chunk.data = newData;
+  }
+  controller.enqueue(chunk);
+}
+
+let rcount = 0;
+function decodeFunction(chunk, controller) {
+  if (rcount++ < 30) { // dump the first 30 packets
+    dump(chunk, 'recv');
+  }
+  const view = new DataView(chunk.data);
+  const checksum = chunk.data.byteLength > 4 ? view.getUint32(chunk.data.byteLength - 4) : false;
+  if (currentCryptoKey) {
+    if (checksum !== 0xDEADBEEF) {
+      console.log('Corrupted frame received, checksum ' +
+                  checksum.toString(16));
+      return; // This can happen when the key is set and there is an unencrypted frame in-flight.
+    }
+    const keyIdentifier = view.getUint8(chunk.data.byteLength - 5);
+    if (keyIdentifier !== currentKeyIdentifier) {
+      console.log(`Key identifier mismatch, got ${keyIdentifier} expected ${currentKeyIdentifier}.`);
+      return;
+    }
+
+    const newData = new ArrayBuffer(chunk.data.byteLength - 5);
+    const newView = new DataView(newData);
+    const cryptoOffset = useCryptoOffset? frameTypeToCryptoOffset[chunk.type] : 0;
+
+    for (let i = 0; i < cryptoOffset; ++i) {
+      newView.setInt8(i, view.getInt8(i));
+    }
+    for (let i = cryptoOffset; i < chunk.data.byteLength - 5; ++i) {
+      const keyByte = currentCryptoKey.charCodeAt(i % currentCryptoKey.length);
+      newView.setInt8(i, view.getInt8(i) ^ keyByte);
+    }
+    chunk.data = newData;
+  } else if (checksum === 0xDEADBEEF) {
+    return; // encrypted in-flight frame but we already forgot about the key.
+  }
+  controller.enqueue(chunk);
+}
+
+onmessage = async (event) => {
+  const {type} = event.data;
+  if (['encode', 'decode'].includes(type)) {
+    const {readableStream, writableStream} = event.data;
+    const transformStream = new TransformStream({
```

Comment:
I think it's more readable if you don't do the double-thing and just do

if (encode) 
  transformStream: encodeFunction
  pipe
} else if (decode)
  transformStream: decodeFunction
  pipe
} else if (setCryptoKey)
  set crypto key
}